net/interfaces/interfaces_linux.go

```diff
@@ -136,13 +137,46 @@ func DefaultRouteInterface() (string, error) {
 
 var zeroRouteBytes = []byte("00000000")
 
+var procIterReadOnce struct {
+	sync.Once
+	v bool
+}
+
+// canIterativelyReadProcFiles reports whether /proc files can be read iteratively.
+// At least on Cloud Run under gVisor, we cannot.
+// See https://github.com/tailscale/tailscale/issues/707#issuecomment-806375600
+func canIterativelyReadProcFiles() bool {
+	procIterReadOnce.Do(func() {
+		f, err := os.Open("/proc/net/route")
+		if err != nil {
+			return
+		}
+		defer f.Close()
+		var buf [1]byte
```

Comment:
It is even more subtle: reading multiple bytes within a line works. This function returns true, that iteratively reading the file works.

If I change this to `var buf [128]byte` to read a full line, this function returns false. The file is 384 bytes long and should contain two more lines, but the read returns 0.

With `var buf [128]byte` defaultRouteInterfaceProcNet() is able to return eth1 as the interface (which is correct), instead of giving up and returning "".

strace I did on a Cloud Run image earlier, reading 128 bytes versus 512 bytes:
```
openat(AT_FDCWD, "/proc/net/route", O_RDONLY|O_CLOEXEC) = 3
read(3, "Iface\tDestination\tGateway\tFlags\t"..., 128) = 128
read(3, "", 128) = 0

openat(AT_FDCWD, "/proc/net/route", O_RDONLY|O_CLOEXEC) = 7
read(7, "Iface\tDestination\tGateway\tFlags\t"..., 512) = 384
read(7, "", 128) = 0
```

Reply:
Wow. I'd file a gVisor bug too. :)

Reply:
Maybe we should just have an is-on-gvisor heuristic.

Something in /proc or uname we could use to detect that cheaply and reliably?

Reply:
Filed https://github.com/google/gvisor/issues/5732 with a minimal test case.

As for identifying Cloud Run:
+ the date string of the kernel it pretends to be might be constant.
+ I don't see anything in /proc/cpuinfo or /proc/meminfo, and it doesn't have other identifying /proc files that I can see

uname -a: `Linux localhost 4.4.0 #1 SMP Sun Jan 10 15:06:54 PST 2016 x86_64 Linux`

/proc/version: `Linux version 4.4.0 #1 SMP Sun Jan 10 15:06:54 PST 2016`

Environment:
```
S2A_ACCESS_TOKEN=long_hex_string
K_REVISION=mycontainer-00112-zom
SHLVL=1
PORT=8080
HOME=/home
K_CONFIGURATION=mycontainer
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
K_SERVICE=mycontainer
```

/proc/meminfo:
```
MemTotal: 2097152 kB
MemFree: 2093604 kB
MemAvailable: 2093604 kB
Buffers: 0 kB
Cached: 1016 kB
SwapCache: 0 kB
Active: 3012 kB
Inactive: 508 kB
Active(anon): 2504 kB
```

/proc/cpuinfo:
```
processor : 0
vendor_id : GenuineIntel
cpu family : 6
model : 85
model name : unknown
stepping : unknown
cpu MHz : 2649.293
fpu : yes
fpu_exception : yes
cpuid level : 13
wp : yes
flags : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves pku
bogomips : 2649.29
bogomips : 2649.29
clflush size : 64
cache_alignment : 64
address sizes : 46 bits physical, 48 bits virtual
power management:

processor : 1
vendor_id : GenuineIntel
cpu family : 6
model : 85
model name : unknown
stepping : unknown
cpu MHz : 2649.293
fpu : yes
fpu_exception : yes
cpuid level : 13
wp : yes
flags : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves pku
bogomips : 2649.29
clflush size : 64
cache_alignment : 64
address sizes : 46 bits physical, 48 bits virtual
power management:
```

Reply:
We discourage explicit "is this gVisor" checks (usually that would be to work around missing functionality, in which case it is better to check for the functionality instead. This case is trickier since it is a weird bug). If you are going to do a check, I recommend using dmesg (aka syslog(2)).

Reply:
Yeah, I really didn't want to add a gVisor check for the same reasons User-Agent sniffing is terrible. Will revert as soon as we can.

Reply:
One way I could think of checking this would be to read as normal, but if you ever competely fill the read buffer and then receive EOF on the next call, then retry from the beginning with a larger buffer.

Unfortunately, I suspect that is kinda difficult to build on bufio.

Reply:
The code in this PR does work around the issue so long as `var buf [128]byte` is used.

Alternately, we might have it try again with a larger buffer if it makes it to the end of the routine and hasn't found the default route. It is weird for a system to not have a default route.